File: solutions/ocaml/collatz-conjecture/1/collatz_conjecture.ml
let rec collatz_conjecture n = 
  if n < 1 then Error "Only positive integers are allowed" else 
    if n == 1 then Ok 0 else
      if n mod 2 == 0 then match (collatz_conjecture (n / 2)) with
      | Ok x -> Ok (x + 1)
      | Error e -> Error e
      else
        match collatz_conjecture (3 * n + 1) with
      | Ok x -> Ok (x + 1)
      | Error e -> Error e

File: solutions/typescript/collatz-conjecture/1/collatz-conjecture.ts
export function steps(count: number): number {
  if (!Number.isInteger(count) || count <= 0) {
    throw 'Only positive integers are allowed'
  }
  if (count == 1) {
    return 0
  } else if(count % 2 == 0) {
    return steps(count / 2) + 1
  } else {
    return steps(3 * count + 1) + 1
  }
}
